### PR TITLE
Presets: shorter index.json file, added option groups, preset priority, updated search, source versioning

### PR DESCRIPTION
**Pull request by limonspb**
please review this PR and merge together-simultaneously with:
https://github.com/betaflight/firmware-presets/pull/163

Configurator TEST binaries [HERE](https://dev.azure.com/Betaflight/Betaflight%20Nightlies/_build/results?buildId=5611&view=artifacts&pathAsName=false&type=publishedArtifacts)
**To test this PR, please, add this preset custom source:**
url: `https://github.com/limonspb/firmware-presets/`
branch: `option_group`
image

This PR makes a couple of fixes and a couple of preset modifications that will help us to maintain presets in the future.
Some of these changes are made out of users' feedback.

1. Adding preset option groups. Some crazy :) people making presets with 15+ options sort of abusing the system. With all them being in a single list it makes it pretty hard on the end user. This PR allows authors to group #$ options:
image

2. Adding presets search priotiry. This moves some "garbage" or "service" presets like defaults etc lower than the useful presets like real tune presets, rc link etc. Every preset can have attribute #$ PRIORITY: value. By default TUNE gets the highest priority. Some categories also have higher priority by default.

3. Cleaning index.json - it was a mistake to include pretty much all the preset attributes to index.json. It makes index.json grow fast and makes search slow. With the new update index.json has ONLY preset attributes that are search related: TITLE, AUTHOR, FIRMWARE_VERSION, PRIORITY, STATUS, KEYWORDS, PATH. With this change configurator actually has to parse all other attributes by itself.

4. Versioning for preset sources. We will need for the future for sure. With the new update every preset source has it's own version. Configurator checks if the version does not match with the expected version it prompts a user:
image

5. Minor style updates and placeholders updates. Making category and firmware version searchable with the text string.

### CODE REVIEW

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by ctzsnooze** [comment]
Really nice!

Being able to provide titles, and advice, for specific option groups will be very helpful.

The new default behaviour (4.3 selected by default, all others empty) and the topmost column field re-naming both are more intuitive.

Being able to go `4.3 whoop` in the search string is really useful.  It means we can, and should remove the firmware version from preset titles and keywords.

**Review by ctzsnooze — DISMISSED** [review]

File: src/tabs/presets/DetailedDialog/PresetsDetailedDialog.js
Comment (haslinghuis) [inline src/tabs/presets/DetailedDialog/PresetsDetailedDialog.js:167]: Can we have an OTHER category for this type?

File: src/tabs/presets/PresetsRepoIndexed/PresetParcer.js
Comment (haslinghuis) [inline src/tabs/presets/PresetsRepoIndexed/PresetParcer.js:3]:
```suggestion
class PresetParser {
```

File: src/tabs/presets/PresetsRepoIndexed/PresetParcer.js
Comment (haslinghuis) [inline src/tabs/presets/PresetsRepoIndexed/PresetParcer.js:31]:
```suggestion
    _parseAttributeLine(preset, line, propertiesMetadata) {
```

File: src/tabs/presets/PresetsRepoIndexed/PresetParcer.js
Comment (haslinghuis) [inline src/tabs/presets/PresetsRepoIndexed/PresetParcer.js:51]:
```suggestion
    _parseOptionDirective(preset, line) {
```

File: src/tabs/presets/PresetsRepoIndexed/PresetParcer.js
Comment (haslinghuis) [inline src/tabs/presets/PresetsRepoIndexed/PresetParcer.js:99]:
```suggestion
    _parseProperty(preset, line, propertyName) {
```

File: src/tabs/presets/PresetsRepoIndexed/PresetParcer.js
Comment (haslinghuis) [inline src/tabs/presets/PresetsRepoIndexed/PresetParcer.js:224]:
```suggestion
PresetParser._sRegExpInclude = /^#\$[ ]+?INCLUDE:[ ]+?(?<filePath>\S+$)/;
```

File: src/tabs/presets/PresetsRepoIndexed/PresetsRepoIndexed.js
Comment (haslinghuis) [inline src/tabs/presets/PresetsRepoIndexed/PresetsRepoIndexed.js:32]:
```suggestion
    _parseInclude(strings, includeRowIndexes, promises)
```

File: src/tabs/presets/PresetsRepoIndexed/PresetsRepoIndexed.js
Comment (haslinghuis) [inline src/tabs/presets/PresetsRepoIndexed/PresetsRepoIndexed.js:35]:
```suggestion
            const match = PresetParser._sRegExpInclude.exec(strings[i]);
```

File: src/tabs/presets/presets.js
Comment (haslinghuis) [inline src/tabs/presets/presets.js:506]:
```suggestion
    if (searchParams.searchString) {
```

**Review by haslinghuis — CHANGES_REQUESTED** [review]
Some minor issues, have tested and works.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

File: src/tabs/presets/DetailedDialog/PresetsDetailedDialog.js
Comment (limonspb) [reply src/tabs/presets/DetailedDialog/PresetsDetailedDialog.js:167]: not sure what you mean here?

Comment (haslinghuis) [reply src/tabs/presets/DetailedDialog/PresetsDetailedDialog.js:167]: image

Comment (limonspb) [reply src/tabs/presets/DetailedDialog/PresetsDetailedDialog.js:167]: oh yeah i see. I figured we will give the authors the flexibility. If they want OTHER group - they can create it. Or they can leave checkboxes without group. They even can even do something like that if thats what they want:
image

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

File: src/tabs/presets/DetailedDialog/PresetsDetailedDialog.js
Comment (ctzsnooze) [reply src/tabs/presets/DetailedDialog/PresetsDetailedDialog.js:167]: Yes I figured that once the PR is merged, authors with lots of options would organise them under 'headings' to make it less of a mess.

Comment (haslinghuis) [reply src/tabs/presets/DetailedDialog/PresetsDetailedDialog.js:167]: Awesome!

File: src/tabs/presets/presets.js
Comment (chmelevskij) [inline src/tabs/presets/presets.js:329]: Why this is a promise? I don'st see anything here what needs it to wrap it in the promise.

File: src/tabs/presets/presets.js
Comment (chmelevskij) [inline src/tabs/presets/presets.js:297]: This is a sync operation, why is it wrapped  inpromise and `then`? What is value in that?

File: src/tabs/presets/PresetsRepoIndexed/PresetParser.js
Comment (chmelevskij) [inline src/tabs/presets/PresetsRepoIndexed/PresetParser.js:3]: A bit of nitpick, but why does every method modify the passed in arguments? This will lead to maintenance hell for other people.

Also unit testing would be much easier if it's just data in data out.

File: src/tabs/presets/presets.js
Comment (limonspb) [reply src/tabs/presets/presets.js:329]: `GUI.showYesNoDialog(dialogSettings)` is not synchronous operation and work on callbacks.
`checkPresetSourceVersion` is a function that is nice to call with .then chaininig.
So I could make this function work with callbacks, but wrapping with promises is nicer here.

File: src/tabs/presets/presets.js
Comment (limonspb) [reply src/tabs/presets/presets.js:297]: not really sync because of the `GUI.showYesNoDialog(dialogSettings)` inside

File: src/tabs/presets/PresetsRepoIndexed/PresetParser.js
Comment (limonspb) [reply src/tabs/presets/PresetsRepoIndexed/PresetParser.js:3]: There are only 4 public methods in this "class":
`constructor(settings)`
`readPresetProperties(preset, strings)`
`removeUncheckedOptions(strings, checkedOptions)`
`isIncludeFound(strings)`

only `readPresetProperties(preset, strings)` actually modifies the argument - `preset`.
"Preset" is a big structure that describes a preset with all its properties/attributes and even CLI text at some point.
This function parses the preset text and sets preset properties according to this text.
Properties like `preset.description`,` preset.warning`, `preset.discussion` etc.

Most of the other "private" functions in this class follow the same "pattern" - they only modify/add properties within `preset`, all other arguments remain untouched.

Comment (chmelevskij) [reply src/tabs/presets/PresetsRepoIndexed/PresetParser.js:3]: But all of those are connected and used only once, in the end reading this code we still need to jump to the methods and see what they actually do to the arguments as they are used once and then just chained. Then there is a mix of mutating and none mutating methods.

Just feels counter intuitive as it needs either knowledge of what it does inside or going and reading every step which modifies the data 🤷‍♂️

**Review by chmelevskij — APPROVED** [review]

File: src/tabs/presets/presets.js
Comment (chmelevskij) [reply src/tabs/presets/presets.js:329]: A, cool, didn't notice that bit. Thanks 😃

File: src/tabs/presets/presets.js
Comment (chmelevskij) [reply src/tabs/presets/presets.js:297]: Gotcha

File: src/tabs/presets/PresetsRepoIndexed/PresetParser.js
Comment (limonspb) [reply src/tabs/presets/PresetsRepoIndexed/PresetParser.js:3]: Yeah, that's not my easiest file to read. But it actually does a heavy job - parsing our own little "programming" language of preset attributes, options, options groups etc.

Comment (chmelevskij) [reply src/tabs/presets/PresetsRepoIndexed/PresetParser.js:3]: Did we define this language anywhere? Might be an interesting task to write proper AST parser instead 😉

Comment (limonspb) [reply src/tabs/presets/PresetsRepoIndexed/PresetParser.js:3]: Yes Sir. The most detailed description is here, made by Ctzsnooze
https://github.com/betaflight/firmware-presets

it is pretty detailed, only missing a couple of most recent  things:
`#$ PRIORITY: ....` and `#$ OPTION_GROUP BEGIN/END`

Comment (limonspb) [reply src/tabs/presets/PresetsRepoIndexed/PresetParser.js:3]: Now the most `full` parser and syntax checker are here:
https://github.com/betaflight/firmware-presets/tree/master/indexer
Here in the Configurator its a simplified version without syntax check and for only necessary attributes.

**Review by haslinghuis — APPROVED** [review]

**Review by ctzsnooze — APPROVED** [review]

**Pull request merged**
